sklearn/metrics/_scorer.py

```diff
@@ -127,6 +127,65 @@ def __init__(self, score_func, sign, kwargs):
         self._score_func = score_func
         self._sign = sign
 
+    @staticmethod
+    def _check_pos_label(pos_label, classes):
+        if pos_label not in list(classes):
+            raise ValueError(
+                f"pos_label={pos_label} is not a valid label: {classes}"
+            )
+
+    def _select_proba(self, y_pred, classes, support_multi_class):
+        """Select the column of `y_pred` when probabilities are provided.
+
+        If `support_multi_class=True` and `y_pred` is valid, then this method
+        is no-op.
+
+        Parameters
+        ----------
+        y_pred : ndarray of shape (n_samples, n_classes)
+            The prediction given by `predict_proba`.
+
+        classes : ndarray of shape (n_classes,)
+            The class labels for the estimator.
+
+        support_multi_class : bool
+            Whether to support multiclass which has been tagged as binary and
+            return `y_pred` as is.
+
+        Returns
+        -------
+        y_pred : ndarray of shape (n_samples,)
+            Probability predictions of the positive class.
+        """
+        if y_pred.shape[1] == 2:
+            pos_label = self._kwargs.get("pos_label", classes[1])
+            self._check_pos_label(pos_label, classes)
+            col_idx = np.flatnonzero(classes == pos_label)[0]
+            return y_pred[:, col_idx]
+
+        err_msg = (
+            f"Got predict_proba of shape {y_pred.shape}, but need "
+            f"classifier with two classes for {self._score_func.__name__} "
+            f"scoring"
+        )
+        if support_multi_class and y_pred.shape[1] == 1:
+            # In _ProbaScorer, `y_true` can be tagged as binary while the
+            # `y_pred` is multi_class. This case is supported when `labels`
+            # is provided in the metric. E.g.:
+            # y_true = np.array([0, 1, 0, 1])
+            # y_score = np.array([[0.1 , 0.8 , 0.1 ],
+            #                     [0.3 , 0.4 , 0.3 ],
+            #                     [0.35, 0.5 , 0.15],
+            #                     [0.  , 0.2 , 0.8 ]])
+            # roc_auc_score(
+            #     y_true, y_score, labels=[0, 1, 2], multi_class='ovo'
+            # )
```

Comment:
I think this is trying to keep the logic from:

https://github.com/scikit-learn/scikit-learn/blob/0a5af0d2a11c64d59381110f3967acbe7d88a031/sklearn/metrics/_scorer.py#L243

which I added in https://github.com/scikit-learn/scikit-learn/pull/15274. This was added because we can have a binary `y_true` with an estimator trained on > 2 classes.

`y_pred.shape[1] == 1` was use to mean `y_pred` came from a classifier with only one class. The check for the shape of `y_pred` was added here: https://github.com/scikit-learn/scikit-learn/commit/94db3d932c393569be97886642049c8920182a77

Reply:
But in scikit-learn we cannot have a classifier train with a single class, isn't it?

Reply:
It is strange, but it can happen:

```py
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_blobs
import numpy as np

X, y = make_blobs(random_state=0, centers=2)
clf = DecisionTreeClassifier()
clf.fit(X, np.zeros_like(y))

clf.classes_
# array([0])
```

Reply:
Uhm. Then, I am confused with `check_classifiers_one_label` :) I have to check. Anyway, I think that the last change make everything more explicit.

Reply:
Oh I see a `must_pass` in the raises. It would be much more explicit to have tag for that I think: `accept_single_label` for instance.